documentation/design_docs/design/infra-resource.adoc

```diff
@@ -0,0 +1,134 @@
+
+A `StandardInfraTemplate` resource defines an infrastructure definition of admin, router and broker
+components. All supported settings for these components can be tuned through this resource. More
+settings that is desired to be configurable can be added. The resource is versioned in the name, and
+as a property.
+
+
+```
+apiVersion: enmasse.io/v1alpha1
+kind: StandardInfraTemplate
+metadata:
+  name: standard-infra-0.21.2
+  namespace: somewhere
+spec:
+  version: 0.21.2
+  router:
+    replicas: 1
+    linkCapacity: 1000
+    allowDurableSubscriptions: false
+    resources:
+      memory: 512Mi
+  broker:
+    addressFullPolicy: FAIL
+    resources:
+      memory: 512Mi
+      disk: 2Gi
+  admin:
+    resources:
+      memory: 512Mi
+```
+
+A `BrokeredInfraTemplate` resource defines configuration for an admin component, and a single broker
+(with future settings might be replicas for instance). Settings for these components can be tuned
+through this resource.
+
+```
+apiVersion: enmasse.io/v1alpha1
+kind: BrokeredInfraTemplate
+metadata:
+  name: brokered-infra-0.21.2
+  namespace: somewhere
+spec:
+  version: 0.21.2
+  broker:
+    addressFullPolicy: PAGE
+    resources:
+      memory: 2Gi
+      disk: 50Gi
+  admin:
+    resources:
+      memory: 512Mi
+```
+
+At present `AddressSpace` is a logical resource managed by tenants. To allow the service admin to
+manage the system without changing the `AddressSpace` resource, an `Infra` resource type is used to
+represent an instance of the above templates. This paves the way later for allowing multiple
+`AddressSpace` resource to be shared by the same `Infra` resource. To start with, the `Infra`
+resource is created automatically by the `address-space-controller` when an `AddressSpace` is created.
+
+An `Infra` resource is used to declare some infrastructure to be created based on a template
+resource (+ authentication service settings?). A component `infra-controller` will watch for
+resources of this kind and instantiate the necessary components as these resources are created.
+Initially, there might as well be a 1:1 mapping between AddressSpace and Infra, with the distinction
+that tenants can see `AddressSpace` resources, where as only service admins can see `Infra`
+resources.
+
+The `Infra` resource points to a particular `*InfraTemplate` resource, indicating what the
+infrastructure should converge to. If the version changes, the `infra-controller` will be responsible
+for converging the `Infra` instance towards the new version based on the new template.
+
+Here is an example of an `Infra` instance:
+
+```
+apiVersion: enmasse.io/v1alpha1
+kind: Infra
+metadata:
+  name: infra
+  namespace: myinfra
+spec:
+  template:
+    kind: StandardInfraTemplate
+    name: standard-infra-0.21.2
+    version: 0.21.2
+  authenticationService:
+    host: example.com
+    port: 5671
+    caBundle: // Base64 encoded CA cert
+```
+
```

Comment:
How many "Infra" definitions can you have within a single installation of EnMasse,  would all the templates be enumerated in a single "infra" ?

Reply:
You can have as many as is needed to support the number of address spaces needed. I.e. the number of infra and address space instances would be 1:1 at first, and then 1:M when 1 Infra can support multiple AddressSpaces.